sdk/elixir/runtime/main.go

```diff
@@ -0,0 +1,177 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"path"
+
+	"github.com/iancoleman/strcase"
+)
+
+const (
+	ModSourceDirPath     = "/src"
+	sdkSrc               = "/sdk"
+	genDir               = "sdk"
+	schemaPath           = "/schema.json"
+	defaultElixirVersion = "1.16.1-erlang-26.2.2-debian-bookworm-20240130-slim"
+)
+
+func New(
+	// +optional
+	sdkSourceDir *Directory,
+) *ElixirSdk {
+	return &ElixirSdk{SDKSourceDir: sdkSourceDir, RequiredPaths: []string{}}
+}
+
+type ElixirSdk struct {
+	SDKSourceDir  *Directory
+	RequiredPaths []string
+}
+
+func (m *ElixirSdk) ModuleRuntime(
+	ctx context.Context,
+	modSource *ModuleSource,
+	introspectionJson string,
+) (*Container, error) {
+	ctr, err := m.CodegenBase(ctx, modSource, introspectionJson)
+	if err != nil {
+		return nil, err
+	}
+
+	modName, err := modSource.ModuleName(ctx)
+	if err != nil {
+		return nil, err
+	}
+	subPath, err := modSource.SourceSubpath(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	mod := normalizeModName(modName)
+	entrypoint := path.Join(ModSourceDirPath, subPath, mod)
+
+	return ctr.
+		WithEntrypoint([]string{
+			"mix", "cmd",
+			"--cd", entrypoint,
+			"mix do deps.get + dagger.invoke",
+		}), nil
+}
+
+func (m *ElixirSdk) Codegen(
+	ctx context.Context,
+	modSource *ModuleSource,
+	introspectionJson string,
+) (*GeneratedCode, error) {
+	ctr, err := m.CodegenBase(ctx, modSource, introspectionJson)
+	if err != nil {
+		return nil, fmt.Errorf("could not load module config: %v", err)
+	}
+
+	return dag.GeneratedCode(ctr.Directory(ModSourceDirPath)).
+		WithVCSGeneratedPaths([]string{genDir + "/**"}).
+		WithVCSIgnoredPaths([]string{"dagger"}), nil
+}
+
+func (m *ElixirSdk) CodegenBase(
+	ctx context.Context,
+	modSource *ModuleSource,
+	introspectionJson string,
+) (*Container, error) {
+	modName, err := modSource.ModuleName(ctx)
+	if err != nil {
+		return nil, err
+	}
+	subPath, err := modSource.SourceSubpath(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	mod := normalizeModName(modName)
+
+	codegenDepsCache, codegenBuildCache := mixProjectCaches(dag, "dagger-codegen")
```

Comment:
There's no need to pass `dag` as it's a globally available client instance. See https://dagger.io/blog/simpler-reusability-with-the-global-client

You can see it in this module's `dagger.gen.go` file.

Reply:
That's my habit that in Elixir there's no global state. 😂 Let me fix.

Reply:
Fixed  in https://github.com/dagger/dagger/pull/6967/commits/a644d96ed3a67dbc2dac46ad64bfd39bcaa32acf

Reply:
> That's my habit that in Elixir there's no global variable. 😂 Let me fix.

Just curious but is that because of how Elixir works or just not implemented yet in the Elixir SDK? 🙂

Reply:
> > That's my habit that in Elixir there's no global variable. 😂 Let me fix.
> 
> Just curious but is that because of how Elixir works or just not implemented yet in the Elixir SDK? 🙂

It's by design. If we want something like Go does, we need to create a new Process, a lightweight thread in Erlang/Elixir, like a goroutine and maintain a client inside it, maybe it need to clone the API interface to talk with the process. That's looks complex.